src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -220,6 +341,18 @@ internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX50
                     throw;
                 }
             }
+            finally
+            {
+                if (innerContext != null)
+                {
+                    // We allocated new context and we want to cache
+                    if (!cacheSslContext || sslAuthenticationOptions.CertificateContext?.SslContexts == null ||
+                        !sslAuthenticationOptions.CertificateContext.SslContexts.TryAdd(sslAuthenticationOptions.EnabledSslProtocols, innerContext))
```

Comment:
Why are we trying to add to the cache in a finally?  If any of the throws throw we probably don't want to come back to this.  Optional caching seems like it should be at the end of the try.